src/org/rust/lang/inspections/SelfConventionInspection.kt

```diff
@@ -0,0 +1,72 @@
+package org.rust.lang.inspections
+
+import com.intellij.codeInspection.LocalInspectionTool
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import org.rust.lang.core.psi.RustImplMethod
+import org.rust.lang.core.psi.RustVisitor
+
+class SelfConventionInspection : LocalInspectionTool() {
+
+    override fun getGroupDisplayName() = "Rust"
+    override fun getDisplayName() = "Self Convention"
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return object : RustVisitor() {
+            override fun visitImplMethod(o: RustImplMethod) {
+                val convention = SELF_CONVENTIONS.find { o.identifier.text.startsWith(it.prefix) } ?: return
+                val selfType = SelfType.from(o.and, o.mut, o.self)
+                if (selfType !in convention.selfTypes) {
+                    holder.registerProblem(o.self ?: o.identifier, convention)
+                }
+            }
+        }
+    }
+
+    companion object {
+        val SELF_CONVENTIONS = listOf(
+            SelfConvention("as_", listOf(SelfType.REF_SELF, SelfType.REF_MUT_SELF)),
+            SelfConvention("from_", listOf(SelfType.NO_SELF)),
+            SelfConvention("into_", listOf(SelfType.SELF)),
+            SelfConvention("is_", listOf(SelfType.REF_SELF, SelfType.NO_SELF)),
+            SelfConvention("to_", listOf(SelfType.REF_SELF))
+        )
+    }
+}
+
+enum class SelfType {
+    NO_SELF,
+    SELF,
+    REF_SELF,
+    REF_MUT_SELF;
+
+    val description: String
+        get() = when (this) {
+            NO_SELF      -> "no self"
+            SELF         -> "self by value"
+            REF_SELF     -> "self by reference"
+            REF_MUT_SELF -> "self by mutable reference"
+        }
```

Comment:
This can also be implemented as

```Kotlin
enum class SelfType(val description: String) {
    NO_SELF("no self"),
    SELF("self by value"),
    REF_SELF("self by reference"),
    REF_MUT_SELF("self by mutable reference");
}
```